// heapsort/heapsort.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.heapsort = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({"/lib/index.js":[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports["default"] = heapsort;
var floor = Math.floor;

function heapsort(array) {
	var comparator = arguments.length <= 1 || arguments[1] === undefined ? ascendantComparator : arguments[1];

	var count = array.length;
	var end = count - 1;

	heapify(array, comparator);

	while (end > 0) {
		swapElems(array, end, 0);
		end = end - 1;
		siftDown(array, 0, end, comparator);
	}

	return array;
}

function ascendantComparator(a, b) {
	return a > b ? 1 : a < b ? -1 : 0;
}

function heapify(array, comparator) {
	var count = array.length;
	var start = floor((count - 2) / 2);

	while (start >= 0) {
		siftDown(array, start, count - 1, comparator);
		start = start - 1;
	}
}

function siftDown(array, start, end, comparator) {
	var root = start;

	while (root * 2 + 1 <= end) {
		var lChild = root * 2 + 1;
		var rChild = lChild + 1;
		var swap = root;

		if (comparator(array[swap], array[lChild]) < 0) {
			swap = lChild;
		}

		if (rChild <= end && comparator(array[swap], array[rChild]) < 0) {
			swap = rChild;
		}

		if (swap === root) {
			return;
		}

		swapElems(array, root, swap);
		root = swap;
	}
}

function swapElems(array, a, b) {
	var tmp = array[a];
	array[a] = array[b];
	array[b] = tmp;
}
module.exports = exports["default"];

},{}]},{},[])("/lib/index.js")
});
